examples/proxy/lib.rs

```diff
@@ -0,0 +1,93 @@
+//! This example demonstrates how the Proxy/Forward pattern can be
+//! implemented in ink!.
+//!
+//! What the contract does is:
+//!
+//!   * Any call to this contract that does not match a selector
+//!     of it is forwarded to a specified address.
+//!   * The instantiator of the contract can modify this specified
+//!     `forward_to` address at any point.
+//!
+//! Using this pattern it is possible to implement upgradable contracts.
+//!
+//! Note though that the contract to which calls are forwarded still
+//! contains it's own state.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+pub mod proxy {
+    /// A simple proxy contract.
+    #[ink(storage)]
+    pub struct Proxy {
+        /// The `AccountId` of a contract where any call that does not match a
+        /// selector of this contract is forwarded to.
+        forward_to: AccountId,
+        /// The `AccountId` of a privileged account that can update the
+        /// forwarding address. This address is set to the account that
+        /// instantiated this contract.
+        admin: AccountId,
+    }
+
+    impl Proxy {
+        /// Instantiate this contract with an address of the `logic` contract.
+        ///
+        /// Sets the privileged account to the caller. Only this account may
+        /// later changed the `forward_to` address.
+        #[ink(constructor)]
+        pub fn new(forward_to: AccountId) -> Self {
+            Self {
+                admin: Self::env().caller(),
+                forward_to,
+            }
+        }
+
+        /// Changes the `AccountId` of the contract where any call that does
+        /// not match a selector of this contract is forwarded to.
+        #[ink(message)]
+        pub fn change_forward_address(&mut self, new_address: AccountId) {
+            assert_eq!(
+                self.env().caller(),
+                self.admin,
+                "caller {:?} does not have sufficient permissions, only {:?} does",
+                self.env().caller(),
+                self.admin,
+            );
```

Comment:
This will remove the debugging output though right?

Reply:
Which debugging output do you mean exactly?
If our `Error` properly implements `Display` this should be fine upon `unwrap`-ing the `Err`.
In case you mean on-chain debugging output: We should not be concerned about this problem in this PR. I think we should make should from a language design perspective that proper Rust error handling is not inferior to panicking.

Reply:
So you would return `Result<(), String>` then?

Reply:
Could you elaborate on what return type you would then choose? `Result<(), String>`? In that case `String` would also be included in the wasm32 target.

Reply:
Create a simple error type with a `Display` impl that tells you everything like you'd do in a real world production Rust codebase. ;)

Reply:
For Wasm file size improvements you could try making the fields over references to `AccountId` or leaving away `caller` sinc the caller is obviously aware of their own `AccountId`.

Reply:
I've updated the PR.